### ISSUE TITLE
nvjpeg decoder

### ISSUE THREAD

**Issue by totaam** [Xpra-org/xpra#3504]
Same as #2984 but for client side decoding this time.

This will allow us to achieve the best possible single picture latency.

Once implemented, it would make sense to tell the server to prefer `jpeg` over `webp` in all(?) cases.

**Comment by totaam** [comment]
The inspiration for the GPU code above was not from [nvidia's own documentation](https://docs.nvidia.com/cuda/nvjpeg/index.html) or sample code - which fails to tell you that the buffer must be a GPU buffer... and the API doesn't fail if you use a CPU buffer either.
Some good code samples:
[using pycuda and glumpy to draw pytorch GPU tensors to the screen without copying to host memory](https://gist.github.com/victor-shepardson/5b3d3087dc2b4817b9bffdb8e87a57c4)
[pycuda gl API documentation](https://documen.tician.de/pycuda/gl.html)

Still TODO: move GPU device selection and keep the context alive as creating it is very expensive.
(and perhaps re-create it when the window is moved to another monitor?)

**Comment by totaam** [comment]
Testing with:
```shell
XPRA_OPENGL_DRAW_REFRESH=0 XPRA_OPENGL_NVJPEG=1 xpra attach tcp://127.0.0.1:10000/ \
    --no-speaker --no-mmap --opengl=force -d nvjpeg,opengl --quality=10
```
A `glxgears` 4K screen update:
```
23,158 glXMakeCurrent: xid=0x5c00033, context=<OpenGL.raw.GLX._types.LP_struct___GLXcontextRec object at 0x7f5eeceb1ac0>
23,159 decompress_with_device(RGB, 141256 bytes, None)
23,159 got image info: 3840x2123 YUV420
23,181 nvjpegDecode took 21ms
23,182 copy_buffer(<pycuda._driver.DeviceAllocation object at 0x7f5eecf14a60>, 24456960)
23,186 RegisteredBuffer(1, pycuda._driver.graphics_map_flags.WRITE_DISCARD)=<pycuda._driver.RegisteredBuffer object at 0x7f5eecec9db0>
23,187 copying from <pycuda._driver.DeviceAllocation object at 0x7f5eecf14a60> to mapping=<pycuda._driver.RegisteredMapping object at 0x7f5eecf14c40> at 0x7f5daa400000
23,193 paint_nvjpeg(GLXWindowContext(0x5c00033)) img=ImageWrapper(RGB:(0, 0, 3840, 2123, 24):PACKED), updating fbo
23,194 set_alignment(3840, 11520, 'RGB') GL_UNPACK_ROW_LENGTH=0, GL_UNPACK_ALIGNMENT=8
23,195 present_fbo: adding (0, 0, 3840, 2123) to pending paint list (size=0), flush=0, paint_screen=True
23,195 do_present_fbo: painting ((0, 0, 3840, 2123),)
23,195 Presenting FBO on screen
23,196 1.do_gl_show(GLDrawingArea(10, (3840, 2123), None)) swapping buffers now
23,196 gl_show after  77ms took  0ms,  1 updates
23,196 GLDrawingArea(10, (3840, 2123), None).gl_frame_terminator()
23,197 <OpenGL.platform.baseplatform.glBindFramebuffer object at 0x7f5f1c086850>(GL_FRAMEBUFFER (GL_FRAMEBUFFER), 2)
23,197 GLDrawingArea(10, (3840, 2123), None).do_present_fbo() done
23,198 glXMakeCurrent: NULL for xid=0x5c00033
```
That's 40ms to decompress and update the screen at 4K, more than half of which (21ms) is spent decoding. The next biggest item is the RGB buffer copying (10ms).

Would using `YUV420P` subsampling speed things up or slow things down by making us download 3 separate buffers?

**Comment by totaam** [comment]
This is now enabled by default and works quite well!
`jpega` is handled by the decoder itself: it merges an `RGB` image with a `Y` one to re-create the `RGBA` data that the window backings can paint with. This costs an extra memory copy, but the complexity of using shaders just isn't worth it IMO.

---

Possible future enhancements:
* paint grayscale directly (single plane) - meh
* decode to `YUV420P` - meh
* cuda context cleanup: #3515
* would it be possible to decode asynchronously and go back to the main loop whilst `nvjpegDecode` does its thing? (using the piecemeal API instead) - unlikely and the code would be hellish

**Comment by totaam** [comment]
It still leaks GPU memory..
With and without the OpenGL buffer passing.
This happens somewhere in `nvjpegDecode`, the memory leak is gone when commenting out that one call.
The memory is freed **when the window is closed** with the opengl backend, so clearly it is somehow attached not just to our process, but the CUDA or GLX context.
With the plain GTK backend, the memory is only freed **when the client process exits**!?
So who owns this buffer and where is it?

---

Things I have tried:
* `NVJPEG_OUTPUT_UNCHANGED` so it doesn't have to do any CSC step itself (thinking that this might be where the leaked buffer was)
* [decode using decoupled phases](https://docs.nvidia.com/cuda/nvjpeg/index.html#nvjpeg-decode-using-decoupled-phases) - but this fails miserably quite early with `xpra.codecs.nvjpeg.decoder.NVJPEG_Exception: nvjpegStateAttachPinnedBuffer failed: INVALID_PARAMETER` - the idea was that using more API calls would perhaps identify the one that is leaking
* keeping the same `PBO` object for the duration of the OpenGL window instead of doing a `glGenBuffers` + `glDeleteBuffers` for every paint with `nvjpeg`.

---

Things I have not tried:
* different driver versions since some versions were known to have a memory leak,  that's more difficult on Fedora with signed modules: you can no longer just run the NVidia binary installer.
* cooking up a plain C PoC
* skip using a `PBO` and load the pixel data directly into a texture?

---

Other notes:
* [Ampere A100 can no nvjpeg](https://forums.developer.nvidia.com/t/nvjpegs-jpeg-encode-performance-on-ampere-a100/202349) _but there is no hardware engine specifically for it_, and there is no support at all for nvenc
* [questions abour decoupled decoding](https://forums.developer.nvidia.com/t/nvjpeg-a-few-questions-abour-decoupled-decoding/108050):
_For the sake of clarity, the sequence in my code is_

1. `cudaGraphicsGLRegisterBuffer()`
2. `cudaGraphicsMapResources()`
3. `cudaGraphicsResourceGetMappedPointer()`
4. `nvjpegDecodeJpegHost()`
5. `nvjpegDecodeJpegTransferToDevice()`
6. `nvjpegDecodeJpegDevice()`
7. (optional)`cuda_kernel()`
8. `cudaGraphicsUnmapResources()`
9. `glBindBuffer(GL_PIXEL_UNPACK_BUFFER, buffer)`
10. `glTexImage2D()`
11. `glBindBuffer(GL_PIXEL_UNPACK_BUFFER, 0)`
12. `glGenerateMipmap()`
This workflow is almost identical to what I have implemented for OpenGL rendering.

**Comment by totaam** [comment]
Perhaps try with [nvjpeg-python](https://github.com/UsingNet/nvjpeg-python), only temporarily as I really don't like the dependencies and the interface.

**Comment by totaam** [comment]
Could also be a vbo leak?
https://github.com/mcfletch/pyopengl/commit/8c82e72d8d2b56d41ee44a393a98a55413c4b39a

**Comment by totaam** [comment]
Or perhaps we need to free the resources / stream whilst the cuda context is active: https://github.com/inducer/pycuda/issues/370#issuecomment-1214833223

**Comment by totaam** [comment]
Testing with b9c24fa099679236d957b30b59be94ac8e0447a1 and a patched pycuda (2be70370a88c371affb26db37b16abf693b8eae6), I do see the new context warnings if I somehow exit the context without freeing the stream. (ie: from an exception)
Unfortunately, when synchronizing properly the leak is still present.

I also tried moving the `pbo` allocation.

**Comment by totaam** [comment]
Found memory allocation related issues in the release notes:
* [nvjpeg 11.5.1](https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#nvjpeg-11.5.1) : _Fixed the issue in which `nvcuvid()` released uncompressed frames causing a memory leak._
* [nvjpeg 11.8.0](https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#nvjpeg-11.8.0) : _Fallback to cudaMalloc/cudaFree when stream ordered allocators are unavailable._

Perhaps we can just try again.

**Comment by totaam** [comment]
As per the commit above, this no longer leaks memory with the latest drivers (at least on MS Windows) and is now enabled by default if the driver version is 522.6 or later.

If we find that the fix is also included in older versions of the drivers, we can change the number used in the version check.

**Comment by dev0x13** [comment]
Hi @totaam. May I ask why do you you rely on the driver version instead of CUDA version for checking? I'm trying to employ nvJPEG in other software, and I need a robust environment check to prevent this severe memory leak in deployments too. I've managed to find only [this one](https://github.com/pytorch/vision/issues/4378) reported issue on this bug in nvJPEG, and the PyTorch maintainers seem to rely on CUDA version for the check. I'm just trying to figure whether it's driver or CUDA is the primary factor of the memory leak presence.
Thank you in advance!

**Comment by totaam** [comment]
@dev0x13 `libnvjpeg.so` is updated with the drivers - not with cuda.
That said, the bug could still be in cuda - I don't know. Someone should ask nvidia.

**Comment by dev0x13** [comment]
@totaam That makes sense, thank you!

**Comment by totaam** [comment]
FWIW: The leak has also been fixed with the Linux drivers - `525.89.02` and potentially earlier.
